File: frontend/src/pages/onboarding/wizardSteps.ts
import type { CandidateStep } from '@/types';

/** Wizard step keys — a superset of the server's `current_step` values. */
export type WizardStepKey =
  | 'REGISTRATION'
  | 'KYC'
  | 'ADDRESS'
  | 'DOCUMENT_CHECKLIST'
  | 'DOCUMENT_UPLOAD'
  | 'VERIFICATION'
  | 'ORIGINAL_VERIFICATION'
  | 'PHOTO'
  | 'LEFT_BIOMETRIC'
  | 'RIGHT_BIOMETRIC'
  | 'DECLARATION'
  | 'SIGNATURE'
  | 'REVIEW';

export interface WizardStep {
  key: WizardStepKey;
  label: string;
  /** Compact label for the narrow-screen stepper. */
  shortLabel: string;
  /**
   * KYC and Address are now captured together with Registration in a single
   * combined step (RegistrationStep), so they no longer appear in the
   * visible stepper on their own. They stay in STEPS (rather than being
   * removed) so goToStep()/stepState() ordering and ReviewStep's "Edit"
   * links back into KYC/Address for corrections keep working.
   */
  hidden?: boolean;
}

export const STEPS: WizardStep[] = [
  { key: 'REGISTRATION', label: 'Registration', shortLabel: 'Register' },
  { key: 'KYC', label: 'KYC', shortLabel: 'KYC', hidden: true },
  { key: 'ADDRESS', label: 'Address', shortLabel: 'Address', hidden: true },
  { key: 'DOCUMENT_CHECKLIST', label: 'Documents', shortLabel: 'Docs' },
  { key: 'DOCUMENT_UPLOAD', label: 'Document Upload', shortLabel: 'Upload', hidden: true },
  { key: 'VERIFICATION', label: 'Document Verification', shortLabel: 'Verify', hidden: true },
  { key: 'ORIGINAL_VERIFICATION', label: 'Original Verification', shortLabel: 'Originals', hidden: true },
  { key: 'PHOTO', label: 'Photo', shortLabel: 'Photo' },
  { key: 'LEFT_BIOMETRIC', label: 'Left Hand Biometric', shortLabel: 'Left Hand' },
  { key: 'RIGHT_BIOMETRIC', label: 'Right Hand Biometric', shortLabel: 'Right Hand' },
  { key: 'DECLARATION', label: 'Declaration', shortLabel: 'Declare' },
  // Signature is removed from the onboarding flow — the key stays (server's
  // current_step enum and candidate_signatures table are untouched) but is
  // hidden so it can never be reached via the stepper or forward navigation.
  { key: 'SIGNATURE', label: 'Signature', shortLabel: 'Sign', hidden: true },
  { key: 'REVIEW', label: 'Final Review', shortLabel: 'Review' },
];

export function stepIndex(key: WizardStepKey): number {
  return STEPS.findIndex((s) => s.key === key);
}

/**
 * The server records the step a candidate has *completed*; the wizard opens on
 * the step that comes next.
 */
export const RESUME_STEP: Record<CandidateStep, WizardStepKey> = {
  // Registration now captures KYC and Address up front, so a candidate
  // still sitting at REGISTRATION (e.g. an old/incomplete draft) resumes
  // straight at Documents rather than the now-hidden KYC/Address steps.
  REGISTRATION: 'DOCUMENT_CHECKLIST',
  KYC: 'DOCUMENT_CHECKLIST',
  ADDRESS: 'DOCUMENT_CHECKLIST',
  // Checklist, Upload, Verification and Original Verification are now one
  // combined Documents step — resuming from any of them (having completed
  // that much server-side) reopens the same merged screen, which shows
  // exactly the sub-sections still outstanding.
  DOCUMENT_CHECKLIST: 'DOCUMENT_CHECKLIST',
  DOCUMENT_UPLOAD: 'DOCUMENT_CHECKLIST',
  VERIFICATION: 'DOCUMENT_CHECKLIST',
  ORIGINAL_VERIFICATION: 'PHOTO',
  PHOTO: 'LEFT_BIOMETRIC',
  LEFT_BIOMETRIC: 'RIGHT_BIOMETRIC',
  RIGHT_BIOMETRIC: 'DECLARATION',
  // Signature is skipped — a candidate whose server-side current_step is
  // still DECLARATION or SIGNATURE (from before this step was removed, or
  // just completed Declaration) resumes straight at Final Review.
  DECLARATION: 'REVIEW',
  SIGNATURE: 'REVIEW',
  SUBMITTED: 'REVIEW',
};

/** Shared props every wizard step receives from the shell. */
export interface WizardStepProps {
  candidateId: string;
  candidate: import('@/types').Candidate;
  /** Re-fetches the candidate so the header/status stays in sync after a save. */
  onSaved: () => Promise<unknown>;
  goNext: () => void;
}

/**
 * Props for RegistrationStep specifically. Unlike every other step, it can
 * run before a candidate exists (candidateId/candidate are null in that
 * "new" mode) — creation happens as part of that step's own submit.
 */
export interface RegistrationStepProps {
  candidateId: string | null;
  candidate: import('@/types').Candidate | null;
  onSaved: () => Promise<unknown>;
  goNext: () => void;
  /** Called after a brand-new candidate is created, with its id. */
  onCreated: (candidateId: number) => void;
}

File: mobile/lib/core/wizard_steps.dart
import '../models/candidate.dart';

/// Canonical wizard step order — mirrors backend
/// src/services/wizardProgressService.js STEP_ORDER exactly, plus the
/// client-only REVIEW step at the end (frontend/src/pages/onboarding/wizardSteps.ts).
enum WizardStepKey {
  registration,
  kyc,
  address,
  documentChecklist,
  documentUpload,
  verification,
  originalVerification,
  photo,
  leftBiometric,
  rightBiometric,
  declaration,
  signature,
  review,
}

class WizardStepMeta {
  final WizardStepKey key;
  final String label;
  final String shortLabel;

  const WizardStepMeta({required this.key, required this.label, required this.shortLabel});
}

const List<WizardStepMeta> kWizardSteps = [
  WizardStepMeta(key: WizardStepKey.registration, label: 'Registration', shortLabel: 'Register'),
  WizardStepMeta(key: WizardStepKey.kyc, label: 'KYC', shortLabel: 'KYC'),
  WizardStepMeta(key: WizardStepKey.address, label: 'Address', shortLabel: 'Address'),
  WizardStepMeta(key: WizardStepKey.documentChecklist, label: 'Documents', shortLabel: 'Docs'),
  WizardStepMeta(key: WizardStepKey.documentUpload, label: 'Document Upload', shortLabel: 'Upload'),
  WizardStepMeta(key: WizardStepKey.verification, label: 'Document Verification', shortLabel: 'Verify'),
  WizardStepMeta(key: WizardStepKey.originalVerification, label: 'Original Verification', shortLabel: 'Originals'),
  WizardStepMeta(key: WizardStepKey.photo, label: 'Photo', shortLabel: 'Photo'),
  WizardStepMeta(key: WizardStepKey.leftBiometric, label: 'Left Hand Biometric', shortLabel: 'Left Hand'),
  WizardStepMeta(key: WizardStepKey.rightBiometric, label: 'Right Hand Biometric', shortLabel: 'Right Hand'),
  WizardStepMeta(key: WizardStepKey.declaration, label: 'Declaration', shortLabel: 'Declare'),
  WizardStepMeta(key: WizardStepKey.signature, label: 'Signature', shortLabel: 'Sign'),
  WizardStepMeta(key: WizardStepKey.review, label: 'Final Review', shortLabel: 'Review'),
];

int wizardStepIndex(WizardStepKey key) => kWizardSteps.indexWhere((s) => s.key == key);

/// The server records the step a candidate has *completed*; the wizard opens
/// on the step that comes next. Mirrors frontend RESUME_STEP exactly.
WizardStepKey resumeStepFor(CandidateStep step) {
  switch (step) {
    case CandidateStep.registration:
      return WizardStepKey.kyc;
    case CandidateStep.kyc:
      return WizardStepKey.address;
    case CandidateStep.address:
      return WizardStepKey.documentChecklist;
    case CandidateStep.documentChecklist:
      return WizardStepKey.documentUpload;
    case CandidateStep.documentUpload:
      return WizardStepKey.verification;
    case CandidateStep.verification:
      return WizardStepKey.originalVerification;
    case CandidateStep.originalVerification:
      return WizardStepKey.photo;
    case CandidateStep.photo:
      return WizardStepKey.leftBiometric;
    case CandidateStep.leftBiometric:
      return WizardStepKey.rightBiometric;
    case CandidateStep.rightBiometric:
      return WizardStepKey.declaration;
    case CandidateStep.declaration:
      return WizardStepKey.signature;
    case CandidateStep.signature:
      return WizardStepKey.review;
    case CandidateStep.submitted:
      return WizardStepKey.review;
  }
}
